PSR 15

ADmad opened the pull request:
This implements PSR 15 support while maintaining the ability to use existing invokable/callable middlewares.

A notable change is you can no longer pass a response instance to `Server::run()` (though I am not sure it has much practical use). If we want to retain the ability to pass a custom response instance to app than we can do by adding a new argument to `BaseApplication`'s constructor.

P.S. I haven't added/updated the docblocks so please avoid commenting on that for now.

ADmad committed:
Update MiddlewareQueue to use MiddlewareInterface instances instead of callables.

It's resolver also wraps callables with DoublePassMiddleware decorator for compatibility
with old callable middlewares.

ADmad committed:
Update Runner to implement RequestHandlerInterface.

ADmad committed:
Update HttpApplicationInterface to extends PSR 15's MiddlewareInterface.

ADmad committed:
Update Server as per changes to Runner.

garas commented on src/Http/MiddlewareQueue.php:
```diff
@@ -34,11 +36,11 @@ class MiddlewareQueue implements Countable
+     * @var \Psr\Http\Server\MiddlewareInterface[]
      */
-    protected $callables = [];
+    protected $middlewares = [];
```
I learned this in my PR https://github.com/cakephp/app/pull/630:

> Plural of middleware is middleware.

ADmad replied:
I think it's acceptable to ignore grammatical rules for a variable name to make it apparent that this is a list of multiple middleware.

Also there are some [source](https://en.wiktionary.org/wiki/middleware) which state that in technical context the plural form is `middlewares`.

ADmad commented:
BC issue can occur if any existing double pass middleware manipulates the response instance before passing it to `$next`. The explanation for it is provided at the bottom [here](https://docs.zendframework.com/zend-stratigility/v3/creating-middleware/#double-pass-middleware). Given that usage is considered incorrect I don't think it's a problem, people should just fix their code (I updated such incorrect usage in tests).

garas commented:
Function in `DoublePassMiddleware::decorateHandler()` could check if it gets same unmanipulated `new Response()` object from `process()` and at least warn about manipulated response.

ADmad commented:
Hmm... that's sounds worth trying.

ADmad committed:
Remove unnecessary assignment.

garas commented on src/Http/Middleware/DoublePassMiddleware.php:
```diff
@@ -0,0 +1,58 @@
+            try {
+                return $handler->handle($request);
+            } catch (Exception $e) {
+                return $response;
```
I think `Runner` could require `$fallbackHandler` in constructor as in [Queue-based request handler](https://www.php-fig.org/psr/psr-15/meta/#queue-based-request-handler) from PSR-15 Meta Document.

Currently in both usages, CakePHP adds last item to queue:
* `Server::run()` adds `Application` to queue as last item.
* `RoutingMiddleware` also adds `$next` to queue as last item.
* Even tests uses same technique.

This would allow to avoid exhausted queue and runner will always be able handle request.

ADmad replied:
Nice, my bad for not reading the meta document.

In practice we are using the fallback approach already as you pointed out, just need to formalize the API accordingly.

The queue exhaustion issue was primary an issue in tests only.

ADmad replied:
I chose a simple approach and now the `Runner` returns a prototype response if the queue is exhausted.

ADmad committed:
Return a prototype response if middleware queue is exhaused without returning a response.

ADmad committed:
Callable should not modify exception before calling $next.

ADmad committed:
Remove unneeded variable assignment.

dereuromark commented on src/Http/Middleware/DoublePassMiddleware.php:
```diff
@@ -45,14 +47,14 @@ public function getCallable(): callable
-                return $response;
+        return function ($request, $res) use ($handler, $response) {
+            if ($res !== $response) {
+                throw new Exception('Callable should not modify response instance before calling $next.');
```
could this also be a RuntimeException?

ADmad replied:
This is `Cake\Core\Exception` which is a `RuntimeException` extension.

ADmad committed:
Update RoutingMiddleware tests.

ADmad committed:
Skip premature response modification check.

ADmad commented:
All tests pass now. I can start cleaning things up and updating the core middlewares to PSR 15 standard upon approval.

@garas I tried your idea of checking for checking for prematurely modified response in `DoublePassMiddleware` but it caused `IntegrationTestTrait` tests to fail because of false positives. I searched a bit but couldn't find any such decorator which checks in other frameworks/libs.

ADmad committed:
Update docblocks.

ADmad committed:
Test using CallableMiddleware.

ADmad committed:
Simplify DoublePassMiddleware.

ADmad committed:
Set default value for property.

markstory commented on tests/TestCase/Http/MiddlewareQueueTest.php:
```diff
@@ -82,8 +83,8 @@ public function testGet()
-        $cb = function () {
-        };
+        $cb = new CallableMiddleware(function () {
+        });
```
Requiring a wrapper will break any existing closure based middleware. I like the reflection idea, it allows both kinds of middleware to co-exist and imparts only a minor performance hit for folks using closure based middleware.

markstory commented:
@ADmad Stickler only has the 'old' CakePHP code standard installed. I've not figured a clever way to have both versions installed.

ADmad committed:
Use reflection of avoid explicitly creating CallableMiddleware instance.

ADmad committed:
Rename DoublePassMiddleware to DoublePassDecoratorMiddleware.

ADmad committed:
Rename CallableMiddleware to CallableDecoratorMiddleware.

ADmad committed:
Fix test.

ADmad committed:
Update docblock.

ADmad commented:
Upgrading the existing double pass core middlewares to PSR-15 is the only thing that remains now.

markstory commented on src/Http/MiddlewareQueue.php:
```diff
@@ -82,25 +85,32 @@ protected function resolve(int $index): ?MiddlewareInterface
+        $info = new ReflectionFunction($middleware);
+        if ($info->getNumberOfParameters() > 2) {
+            return $this->middlewares[$index] = new DoublePassDecoratorMiddleware($middleware);
+        }
```
This is good. Makes adding middleware backwards compatible and easier to use in the future as well.

ADmad committed:
Fix CS error

ADmad committed:
Update BodyParserMiddleware to PSR 15 standard.

ADmad committed:
Update CsrfProtectionMiddleware to PSR 15 standard.

ADmad committed:
Update EncryptedCookieMiddleware to PSR 15 standard.

ADmad committed:
Update SecurityHeadersMiddleware to PSR 15 standard.

ADmad committed:
Update RoutingMiddleware to PSR 15 standard.

ADmad committed:
Update RequestHandler to PSR 15 standard.

ADmad committed:
Update LocaleSelectorMiddleware to PSR 15 standard.

ADmad committed:
Update ErrorHandlerMiddleware to PSR 15 standard.

ADmad committed:
Fix tests.

ADmad committed:
Make Runner use a fallback handler if middleware queue doesn't generate a response.

HttpAllicationInterface now extends RequestHanlderInterface instead of MiddlewareInterface.

ADmad commented on src/Routing/Middleware/RoutingMiddleware.php:
```diff
@@ -148,15 +147,17 @@ public function __invoke(
+
+        $matching[] = new CallableDecoratorMiddleware(function ($req, $handl) use ($handler) {
+            return $handler->handle($req);
+        });
```
@garas Done

ADmad committed:
Fix CS error.

ADmad commented:
All changes done as per feedback. This is ready for final review.

ADmad committed:
Tweak exception handling code.

ADmad committed:
Show useful message for dev.

ADmad commented on tests/TestCase/Http/RunnerTest.php:
```diff
@@ -53,36 +57,38 @@ public function setUp()
     public function testRunResponseReplace()
     {
+        $this->markTestSkipped(
+            'Skip until we figure out a way to test for premature response modification in DoublePassMiddleware.'
```
I couldn't find a way to check and warn dev for this, so I think the test should be just dropped. The middleware example in [manual](https://book.cakephp.org/3.0/en/controllers/middleware.html#creating-middleware) does have a comment stating that response should be modified *after* calling $next.

ADmad commented:
@markstory Updates done as per your feedback.

ADmad committed:
Remove cruft of defunct dipsatcher filters.

ADmad committed:
Merge branch '4.x' into psr-15

ADmad committed:
Remove test for unimplementable feature.

ADmad committed:
Use interface instance of concrete class for type checks.

ADmad committed:
Fix CS errors.

ADmad commented:
@markstory I cleanup up `ActionDispatcher`.

markstory reviewed (approved):
Looks good to me.
